Simplify codebase: replace ConstantToClassConstantBase with configurable rule

bbrala opened the pull request:
## Description
The codebase can be simplified a lot by using configururable rector rules. This change replaces all rules based on `ConstantToClassConstantBase` to a new rule `ConstantToClassConstantRule`. This means we can remove a lot of classes.

Also added a new option to the testbase so we can send actual configuration to the testcases.

## To Test
- There is a new tests added that uses the testbase from one of the removed classes.

## Drupal.org issue
https://www.drupal.org/project/rector/issues/3372902

bbrala committed:
create ConstantToClassConstantRector

bbrala committed:
create test for ConstantToClassConstantRector based on DatetimeStorageTimezoneRector

bbrala committed:
rewrite deprecations to ConstantToClassConstantRector

bbrala committed:
Remove classes that are superceeded by configurable rector rule

bbrala committed:
ConstantToClassConstantRector should use its own constants.

bbrala committed:
Move test since there was a typo

bbrala committed:
Dont use array shape for config

bbrala committed:
Use ConfiguredCodeSample for the RuleDefinitioin

bbrala committed:
Make sure we don't change how other tests are run because of new argument in DeprecationBase

mglaman commented on tests/src/Rector/Deprecation/DeprecationBase.php:
```diff
@@ -18,14 +18,17 @@ class DeprecationBase {
-        else {
+    public static function addClass(string $rectorClass, RectorConfig $rectorConfig, bool $add_notice_config = TRUE, array $configuration = []) {
+        if ($add_notice_config) {
+            $configuration = ['drupal_rector_notices_as_comments' => '%drupal_rector_notices_as_comments%'] + $configuration;
```
nit: just add the key versus merge.

```
$configuration['drupal_rector_notices_as_comments'] = '%drupal_rector_notices_as_comments%';
```

mglaman commented on src/Rector/Deprecation/ConstantToClassConstantRector.php:
```diff
@@ -0,0 +1,108 @@
+        $this->deprecatedConstant = $configuration[static::DEPRECATED_CONSTANT];
+        $this->constantFullyQualifiedClassName = $configuration[static::CONSTANT_FULLY_QUALIFIED_CLASS_NAME];
+        $this->constant = $configuration[static::CONSTANT];
+    }
```
We should add validation here, which may help debug. Apply `?? ''` to each and then check if any property is `=== ''` and throw an exception that it must be properly configured.

bbrala committed:
Apply feedback

bbrala committed:
Use a ValueObject for configurable rector rule

mglaman commented on src/Rector/Deprecation/ConstantToClassConstantRector.php:
```diff
@@ -2,9 +2,12 @@
 use Rector\Core\Contract\Rector\ConfigurableRectorInterface;
 use Rector\Core\Rector\AbstractRector;
+use Rector\Renaming\Contract\MethodCallRenameInterface;
+use RectorPrefix202304\Webmozart\Assert\Assert;
```
This `Assert` isn't available as a dependency. So we'd have to add it, or see if there is a non-namespaced namespace.

bbrala replied:
Fixed

bbrala committed:
Remove Assert and just foreach check

agentrickard committed:
Merge in changes from bbrala:feature/remove-nodetoaddcolector

agentrickard committed:
Simple fix for PARENT_NODE

agentrickard committed:
Simple fixes for PARENT_NODE

agentrickard committed:
Update phpstan.neon with note

agentrickard committed:
Fix typo

agentrickard committed:
Fix ussage of NEXT_NODE

agentrickard committed:
Merge branch 'main' into stan-fails

bbrala committed:
Whitespace (and trigger ci)

bbrala committed:
Fix find parent by type by cheating the method into a trait

bbrala committed:
Use the findParentTypeTrait to fix DrupalSetMessageRector

bbrala committed:
Update ConstantToClass.php

bbrala committed:
rename ValueObject to ConstantToClassConfiguration

agentrickard committed:
Push potential changes for review -- even though failing

agentrickard committed:
Merge branch 'stan-fails' into fix/find-parent-by-type

agentrickard committed:
Merge pull request #227 from bbrala/fix/find-parent-by-type

Fix/find parent by type

agentrickard committed:
Update phpunit config

mglaman committed:
listen for expression if needing to add comment

agentrickard committed:
Merge pull request #230 from mglaman/parent-traverse-fix

listen for expression if needing to add comment

agentrickard committed:
Fixes phpstan errors

bbrala committed:
ExtensionPathBase needs to use the expression node to be able to add comments.

bbrala committed:
Add a CommentingVisitor

Adds comments to statements if there is an attribute on the node set at some other point. This works around having to manipulate parents to get comments. Next iteration would be to remove the need for knowing a parent and setting an attribute on the childnodes 'parent statement' or something similar.

bbrala committed:
Set correct attribute

bbrala committed:
Revert "listen for expression if needing to add comment"

This reverts commit 75ae4098dc1eb795e05866831bfb83fb2f45899e.

bbrala committed:
Be more specific

bbrala committed:
HAve to update to use interface also in functional tests that do not use dev dependency

bbrala committed:
Refactor DBBase to use expression

bbrala committed:
Some small fixes to green last few unit tests

bbrala committed:
refactor LinkGeneratorTraitLRector to expression

bbrala committed:
refactor LinkGeneratorTraitLRector to expression

bbrala committed:
DBBase fix for assigned expressions

bbrala committed:
DBBase fix logic

bbrala committed:
refaactor EntityInterfaceLinkRector to use stmt

bbrala committed:
Refactor EntityInterfaceLinkRector to use stmt and also consider Assign calls

bbrala committed:
Refactor EntityInterfaceLinkRector to use stmt and also consider Assign calls

bbrala committed:
Update rector.php whitespace

bbrala committed:
Refactor DBBase to allow for chained methods

bbrala committed:
EntityLoadBase to stmt is too hard

bbrala committed:
EntityLoadBase to stmt is too hard, updating expected result

bbrala committed:
EntityLoadBase to stmt is too hard, updating expected result
